cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1242,27 +1372,45 @@ class S3FileSystem::Impl {
   // Recursive workhorse for GetTargetStats(FileSelector...)
   Status Walk(const FileSelector& select, const std::string& bucket,
               const std::string& key, std::vector<FileInfo>* out) {
-    int32_t nesting_depth = 0;
-    return Walk(select, bucket, key, nesting_depth, out);
-  }
+    bool is_empty = true;
 
-  Status Walk(const FileSelector& select, const std::string& bucket,
-              const std::string& key, int32_t nesting_depth, std::vector<FileInfo>* out) {
-    if (nesting_depth >= kMaxNestingDepth) {
-      return Status::IOError("S3 filesystem tree exceeds maximum nesting depth (",
-                             kMaxNestingDepth, ")");
-    }
+    auto handle_error = [&](const AWSError<S3Errors>& error) -> Status {
+      if (select.allow_not_found && IsNotFound(error)) {
+        return Status::OK();
+      }
+      return ErrorToStatus(std::forward_as_tuple("When listing objects under key '", key,
+                                                 "' in bucket '", bucket, "': "),
+                           error);
+    };
 
-    bool is_empty = true;
-    std::vector<std::string> child_keys;
+    auto handle_recursion = [&](int32_t nesting_depth) -> Result<bool> {
+      if (nesting_depth >= kMaxNestingDepth) {
+        return Status::IOError("S3 filesystem tree exceeds maximum nesting depth (",
+                               kMaxNestingDepth, ")");
+      }
```

Comment:
since this is common to any recursive walk, should it be inlined in ListObjectsV2Handler::DoHandle or so?